src/raspi/component/latch/pin/pin.py

```diff
@@ -0,0 +1,126 @@
+from gpiozero import DigitalOutputDevice
+from state_management import create_masked_context, device_action, output_device_ctx
+
+enab_pin_ctx = create_masked_context(output_device_ctx, "enab_pin")
+data_pin_ctx = create_masked_context(output_device_ctx, "data_pin")
+addr_pin_ctx = create_masked_context(output_device_ctx, "addr_pin")
+
+
+@device_action(addr_pin_ctx)
+def set_addr(dev: DigitalOutputDevice, state: int) -> None:
+    """
+    Set the address pin.
+
+    Args:
+        dev (DigitalOutputDevice): the address pin
+        state (int): 1 to turn the pin on, 0 to turn it off
+    """
+    dev.on() if state > 0 else dev.off()
+
+
+@device_action(addr_pin_ctx)
+def get_addr(dev: DigitalOutputDevice) -> int:
+    """
+    Get the state of the address pin.
+
+    Args:
+        dev (DigitalOutputDevice): the address pin
+
+    Returns:
+        int: 1 if the pin is on, 0 if the pin is off
+    """
+    return dev.value
+
+
+@device_action(addr_pin_ctx)
+def toggle_addr(dev: DigitalOutputDevice) -> int:
+    """
+    Toggle the state of the address pin.
+
+    Args:
+        dev (DigitalOutputDevice): the address pin
+
+    Returns:
+        int: 1 if the pin is on, 0 if the pin is off
+    """
+    return dev.toggle()
+
+
+@device_action(data_pin_ctx)
+def set_data(dev: DigitalOutputDevice, state: int) -> None:
+    """
+    Set the data pin.
+
+    Args:
+        dev (DigitalOutputDevice): the data pin
+        state (int): 1 to turn the pin on, 0 to turn it off
+    """
+    dev.on() if state > 0 else dev.off()
+
+
+@device_action(data_pin_ctx)
+def get_data(dev: DigitalOutputDevice) -> int:
+    """
+    Get the state of the address pin.
+
+    Args:
+        dev (DigitalOutputDevice): the address pin
+
+    Returns:
+        int: 1 if the pin is on, 0 if the pin is off
+    """
+    return dev.value
+
+
+@device_action(data_pin_ctx)
+def toggle_data(dev: DigitalOutputDevice) -> int:
+    """
+    Toggle the state of the data pin.
+
+    Args:
+        dev (DigitalOutputDevice): the address pin
+
+    Returns:
+        int: 1 if the pin is on, 0 if the pin is off
+    """
+    return dev.toggle()
+
+
+@device_action(enab_pin_ctx)
+def set_enab(dev: DigitalOutputDevice, state: int) -> None:
+    """
+    Set the enable pin.
+
+    Args:
+        dev (DigitalOutputDevice): the enable pin
+        state (int): 1 to turn the pin on, 0 to turn it off
+    """
+    dev.on() if state > 0 else dev.off()
+
+
+@device_action(enab_pin_ctx)
+def get_enab(dev: DigitalOutputDevice) -> int:
+    """
+    Get the state of the address pin.
+
+    Args:
+        dev (DigitalOutputDevice): the address pin
+
+    Returns:
+        int: 1 if the pin is on, 0 if the pin is off
+    """
+    return dev.value
+
+
+@device_action(enab_pin_ctx)
```

Comment:
Yeah so the main difference between the two function is everything that is abstracted in the decorator. 
The main issue I have with merging the two is that the decorator has a type safety embedded within the decorator, so if we pass in an identifier for another component the function will not run and fail immediately. If we merge the two functions and combine the components allowed then we would allow the two components to be swapped and the program will not fail where it should. For function as small as toggle output devices, it's not a huge deal if we have a typesafe environment or not, but since the same decorators will be used for bigger components and that becomes more of a problem. However, the thing about these enable and data pin devices is that they extend from the generic component of output device, so we can technically make a device action function on the generic class and use that, but obviously that loses type safety. If we can somehow define two differently named functions with a single definition then this can be fixed, but due to limitation of the language, we can not. 
This is not a fault of the library, but the more classes we use the more the architecture will fail. This is the reason why I recommend more usage of structs and simple dataclasses in this architecture because every class methods is going to essentially be a private method that will never be exposed to the final set of functions that we can use in the main script file. Its also kind of the only way we can do private and protected methods in python, by having global methods like this architecture. 
So the equivalent of this issue in OOP is like asking to define foo.publicMethod() and bar.publicMethod() only once because they both use self.privateMethod() which is not something we would consider thus I have dismissed this issue and moved on. Similarly to how we can make foo and bar extend the same class and have the extended class have publicMethod() this architecture will let you define publicMethod for the extended parent class like defining publicMethod on the output device class that both enable and data pins are. 
This issue is not an issue in OOP thus I don't consider it to be an issue in this architecture, but to fix this issue in this architecture we would basically be saying that OOP also fails with this issue so I decided to move on.